massa-module-cache/src/controller.rs

```diff
@@ -144,21 +148,25 @@ impl ModuleCache {
                 }
             }
         };
-        Ok(module)
+        Ok((module, execution_gas))
     }
 
-    /// Load a temporary module from arbitrary bytecode
+    /// Load a temporary module from arbitrary bytecode.
+    /// Also checks that the provided execution gas is enough to pay for the instance creation cost.
+    ///
+    /// Returns the module and the remaining gas after compilation.
     pub fn load_tmp_module(
         &self,
         bytecode: &[u8],
         limit: u64,
-    ) -> Result<RuntimeModule, CacheError> {
+    ) -> Result<(RuntimeModule, u64), CacheError> {
         debug!("load_tmp_module");
-        Ok(RuntimeModule::new(
-            bytecode,
-            limit,
-            self.cfg.gas_costs.clone(),
-            Compiler::SP,
-        )?)
+        limit
```

Comment:
shouldn't this be 
`let limit = limit.checked_sub()`
(i think that the gas is not paid anymore here)

Reply:
It is paid, this is only a check, not supposed to actually take anything from the limit. Basically what happens now is:
1. Check that given gas is above max_instance_cost
2. Module compilation & deserialization is always done with max_instance_cost as input, which is supposed to be enough for the lock-out gas used to initialize. And it needs more the module will be considered invalid as we don't want to accept SC with insane instance cost (being above 2.1 mil is crazy). So we don't take limit input for compil & deser.
3. Call execution (run_function & run_main) with given gas, runtime refills the module with the correct amount as you can change the internal metering values after the instance step is done. So; compil payed by ABI costs / OP costs, VM & instance gas amount hard checked by cache, VM & instance + execution payed by runtime

Reply:
Added doc on it

Reply:
I was missing the very last part `VM & instance + execution payed by runtime`. Many thanks for the clarification.